### ISSUE TITLE
Is having DARIA be a single tenant application still serving us?

### ISSUE THREAD

**Issue by colinxfleming** [DARIAEngineering/dcaf_case_management#1817]
TK

**Comment by colinxfleming** [comment]
So here's what this was all about: When we made the decision to spin up new instances for BAF and BRAF, we didn't have a great sense of how much custom work they'd require, or how close their systems were to us. Turns out funds will mostly adhere to our systems pretty well, so that's good, and raises the question: Should we pool everything into one app and scope users to a particular fund?

Some arguments for:

* Hobby dynos are $7 and db instances are like... $15 or something for not that much horsepower (and as @xmunoz has pointed out in the past, rails does way better on a heavier dyno than the heroku hobby). Pooling resources into one environment would both save us money in the long run and let us use more powerful instances
* Less to manage! We'd only have to worry about one production environment, so less of a devops burden
* Onboarding new funds gets easier - don't have to do a devops song and dance, and would all be database-based

Some arguments against:

* This is definitely not free from an engineering perspective; we'd have to potentially rewire quite a bit to scope everything to particular funds
* compromising creds gets nastier - now a breach potentially affects everyone instead of just one fund. (though that is still true if heroku creds get compromised)

**Comment by colinxfleming** [comment]
@lomky @harumhelmy @mdworken @xmunoz interested in your thoughts here!

**Comment by colinxfleming** [comment]
note to self https://github.com/ErwinM/acts_as_tenant

**Comment by lomky** [comment]
I'm not super convinced the positives here are worth the real and potential downsides. 
We save one some one time costs (app spin up) and some literal small costs. I don't really find the multile prod envs that complicated to manage at our current scale.

On the downside, now each of the funds has to trust every other fund in their security practices, and every new fund we bring on increases the risks to existing funds, which creates some negative incentives. 

My gut instinct is that this doesn't feel quite worth it, but I'm interesting in what mitigations we could implement for the downsides, as if we keep growing past a dozen funds the devops costs start to stack up.

**Comment by colinxfleming** [comment]
re: devops costs stacking up, we're starting to see some early examples of that with postgres - a plan with a security bump (standard 0) is $50/m. Spread across our six client funds that's $300/m, which would roughly quadruple our hosting bill at the moment.

**Comment by colinxfleming** [comment]
Nine months since we've bumped this! Here's the state of play:
* We want boofier postgresses for everyone, but buying boofier postgresses is cost-prohibitive on heroku.
* Our costs also go up fairly linearly right now as we add new clients.

* I played around with one option, `acts_as_tenant`, in #2300  - this is a more aggressive approach that consolidates everyone into shared dynos and database. It seems pretty straightforward to implement, though we'd also want to port Line to a model which that does not do in the tryout (and there are ways to hack around it). This approach assumes everyone on a single shared dyno-set, and a single shared database, and funds' particular data is scoped by organization_id or similar. Advantages are that this keeps everything in heroku, lets us get significantly more bang for our buck on heroku; disadvantages are that it's a lot of code changes, that shared infra would mean that more traffic-heavy funds (DCAF, specifically) would potentially cause outages for everyone, and problems we've been able to skirt over (looking at you, n+1 queries) would start to bite us harder.
* @xmunoz proposed a shared database model - we get a single very boofy database from AWS and multitenant that via schemas - every fund gets their own user that can only access their own schema, and every fund keeps their own dyno. Advantages here are that funds stay iso'd and can only break their own instances, and fewer code changes. Disadvantages are that we don't get the cost network effect, database/user config could be potentially tricky to get right.

(Both approaches do not perfectly solve for the points @lomky raises above about security practices and negative incentives, but I think we are in agreement that boofier postgres is worth those tradeoffs.)

Our good friends MAC are taking an approach similar to acts_as_tenant, fwiw.
